### PR TITLE
chore(remap): `merge` returns the result rather than mutates the parameter

### LINKED ISSUE #6265: Update the Remap `merge` function to return a copy of the target map
The Remap `merge` function modifies the `to` map in-place. This deviates in behavior from other functions and is surprising. Instead, it should return a new merged map, leaving the original map untouched.

### PR DESCRIPTION
**Pull request by FungusHumungus**
Closes #6265 

`merge` used to mutate the first parameter rather than return the merged result, so you would have to do:

```
merge(., parse_json(...))
```

This changes the function so it now returns the result and leaves the first parameter untouched, so you have to do:

```
. = merge(., parse_json(...))
```

This brings it inline with how the rest of the Remap functions work.

Note I haven't needed to change much of the documentation as that was assuming merge worked like this anyway!

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/remap-functions/src/merge.rs [hunk 1]
```diff
@@ -121,88 +112,39 @@ mod tests {
...
+                             from: map![ "key2": "val2",
+                                          "child": map! [ "grandchild2": "val2" ] ],
+                             deep: true
+            ],
+            want: Ok( map![ "key1": "val1",
+                             "key2": "val2",
+                             "child": map! [ "grandchild1": "val1",
+                                              "grandchild2": "val2" ] ])
+        }
+    ];
```
Comment (pablosichert) [inline lib/remap-functions/src/merge.rs:149]: Everything within the `test_function!` looks quite weirdly formatted (even the number of spaces for the indentation does not match). Does `cargo fmt` bail in macros?

**Comment by pablosichert** [comment]
Adding some tests that show how this can fail at compile- and runtime would be great. Especially since we are doing some intersection and unions on the types here.

**Review by pablosichert — APPROVED** [review]
Other than that – LGTM

File: lib/remap-functions/src/merge.rs (hunk 1 above)
Comment (JeanMertz) [reply lib/remap-functions/src/merge.rs:149]: Yeah it does, unfortunately. It's something I really dislike as well 😭

It _sometimes_ works, but it seems to depend on the amount of code inside the macro, not sure how that works.

Comment (FungusHumungus) [reply lib/remap-functions/src/merge.rs:149]: > Does cargo fmt bail in macros?

Yes, and even more annoyingly Emacs always wants to auto-indent the second line with an extra space for some reason, so if I'm not extra cautious these issues creep in. I'll fix it.

**Comment by FungusHumungus** [comment]
> Adding some tests that show how this can fail at compile- and runtime would be great. Especially since we are doing some intersection and unions on the types here.

The only way this function can fail is if it is passed parameters that are the wrong type. Providing it receives two maps it cannot error. The type checks are captured by type_def of the function plus the parameters list ensures all non-map parameters error before it gets to this code.

I have some type_def tests to ensure it is fallible if non map parameters are passed in to make sure.

File: lib/remap-functions/src/merge.rs [hunk 2]
```diff
@@ -120,89 +111,79 @@ where
...
+            ],
+            want: Ok(map![ "key1": "val1",
+                           "key2": "val2" ])
+        }
+
+        shallow {
+            args: func_args![ to: map![ "key1": "val1",
+                                        "child": map! [ "grandchild1": "val1" ] ],
+                              from: map![ "key2": "val2",
+                                          "child": map! [ "grandchild2": "val2" ] ]
```
Comment (pablosichert) [inline lib/remap-functions/src/merge.rs:169]: There's one superfluous space.

File: lib/remap-functions/src/merge.rs [hunk 3]
```diff
@@ -120,89 +111,79 @@ where
...
+            want: Ok(map![ "key1": "val1",
+                           "key2": "val2",
+                           "child": map! [ "grandchild2": "val2" ] ])
+        }
+
+        deep {
+            args: func_args![to: map![ "key1": "val1",
+                                       "child": map![ "grandchild1": "val1" ] ],
+                             from: map![ "key2": "val2",
+                                         "child": map![ "grandchild2": "val2" ] ],
```
Comment (pablosichert) [inline lib/remap-functions/src/merge.rs:180]: Here as well.

File: lib/remap-functions/src/merge.rs (hunk 2 above)
Comment (pablosichert) [reply lib/remap-functions/src/merge.rs:169]: Ah, nevermind, I see now that they are just aligned on the left!

**Pull request merged**